geochem_inverse_optimize.py

```diff
@@ -209,16 +214,19 @@ def process_data(data_dir: str, data_filename: str, excluded_elements: Optional[
     except cp.error.SolverError as err:
       print(f"\033[91mSolver Error - skipping this element!\n{err}")
       continue
+        
+    obs,pred=[],[]
+    for sample in element_data.keys():
```

Comment:
This entire loop can and should be replaced with list comprehensions, like so:
```
obs = [element_data[sample] for sample in element_data.keys()]
pred = [predictions[sample] for sample in element_data.keys()]
```
Doing this allows you to eliminate the temporary variables and assign directly to `results`, below.